jerry-ext/include/jerryscript-ext/context.h

```diff
@@ -0,0 +1,84 @@
+/* Copyright JS Foundation and other contributors, http://js.foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef JERRYX_CONTEXT_H
+#define JERRYX_CONTEXT_H
+
+#include "jerryscript.h"
+
+/**
+ * Function to pass to jerry_init_with_user_context() as its second parameter
+ */
+void *jerryx_context_init (void);
+
+/**
+ * Function to pass to jerry_init_with_user_context() as its third parameter
+ */
+void jerryx_context_deinit (void *user_context_p);
+
+/**
+ * Definition of a context slot.
+ */
+typedef struct
+{
+  /** the initialization function */
+  jerry_user_context_init_cb init_cb;
+
+  /** the deinitialization function */
+  jerry_user_context_deinit_cb deinit_cb;
+} jerryx_context_slot_t;
+
+#ifdef __MACH__
+#define JERRYX_CONTEXT_SECTION "__DATA,jerryx_ctx_slots"
+#else /* !__MACH__ */
+#define JERRYX_CONTEXT_SECTION "jerryx_ctx_slots"
+#endif /* __MACH__ */
+
+/**
+ * Define a slot.
+ *
+ * Use this macro at the global level in a source file to define a slot. Afterwards you can use JERRYX_CONTEXT_SLOT
+ * further down in the same file to retrieve the contents of the slot from the user context.
+ *
+ * @param slot_name the suffix to be given to the global static variable holding the slot definition.
+ * @param init_cb_ptr a pointer to a function of type jerry_user_context_init_cb to be called when initializing a slot.
+ * @param deinit_cb_ptr a pointer to a function of type jerry_user_context_deinit_cb to be called when freeing a slot.
+ */
+#define JERRYX_CONTEXT_DEFINE_SLOT(slot_name, init_cb_ptr, deinit_cb_ptr)         \
+static jerryx_context_slot_t __jerryx_context_slot_ ## slot_name                  \
+__attribute__((used, section(JERRYX_CONTEXT_SECTION), aligned(sizeof(void *)))) = \
```

Comment:
The problem is that context.h is public whereas section-utils.h is private.

So, I guess I should publish section-utils.h - maybe as jerryscript-ext/section.h - because being able to easily manipulate sections seems like it might be beneficial.

Who says you can't have a header-only extension, eh? :)